# Epitope-Only Held-Out Preparation Workflow

This notebook prepares the new IEDB epitope table export for the epitope-only held-out workflow. Rows are kept only when the epitope is a linear, unmodified, canonical 9-mer using the 20 standard amino acids. Any peptide name with non-canonical characters such as `+`, `X`, `B`, `Z`, `J`, brackets, or other uncertainty/modification notation is excluded. Rows with populated `Modified Residue(s)` or `Modifications` fields are also excluded before deduplication.

The downstream training notebook is `Epitope_A0301_Heldout_Training_Run.ipynb`, where these epitopes are treated as HLA-A03:01 binders and unique label-0 peptides from `hla_only.txt` are introduced as non-binders. Same-allele negatives are used first, then other-allele negatives are added to create a 1:1 split.

In [2]:
# Recreate all generated CSV/SVG/TXT outputs from the workbook.
%run epitope_table_heldout_run.py --no-notebook


Workbook: epitope_table_export_1779969300.xlsx
Raw epitope rows: 9833
Canonical retained rows before dedupe: 9648
Unique canonical 9-mers: 9648
Default held-out amino acid: R
Training/CV peptides without R: 5523
Evaluation peptides with R: 4125
Network training is handled by Epitope_A0301_Heldout_Training_Run.ipynb after adding balanced label-0 non-binders.


## Current Saved Summary

| Metric | Value |
| --- | --- |
| Raw epitope rows | 9833 |
| Retained canonical 9-mer rows before dedupe | 9648 |
| Unique canonical 9-mers | 9648 |
| Train/CV peptides without R | 5523 |
| Evaluation peptides with R | 4125 |
| Consensus, all canonical 9-mers | RLYPPLLLK |
| Consensus, no R train/CV | KLYPPLLLK |
| Consensus, with R evaluation | RLYRRLLRK |
| Excluded: modified_or_uncertain_annotation | 185 |

## Held-Out Amino-Acid Candidates

The default split is held-out `R` to match the downstream A03:01 training notebook. The candidate table below also shows that the workbook can support other held-out choices; set `DEFAULT_HELD_OUT_AA` in `epitope_table_heldout_run.py` and rerun if you want a different residue.

| Held_Out_AA | Train_Count_No_AA | Evaluation_Count_With_AA | Anchor_P2_or_P9_Count | Anchor_Only_P2_or_P9_Count | Rank_By_Train_Count | Rank_By_Evaluation_Count |
| --- | --- | --- | --- | --- | --- | --- |
| C | 8965 | 683 | 67 | 57 | 1 | 20 |
| W | 8577 | 1071 | 134 | 102 | 2 | 19 |
| M | 7734 | 1914 | 526 | 418 | 3 | 18 |
| H | 7705 | 1943 | 457 | 366 | 4 | 17 |
| D | 7616 | 2032 | 55 | 44 | 5 | 16 |
| E | 7258 | 2390 | 202 | 138 | 6 | 15 |
| N | 7242 | 2406 | 69 | 52 | 7 | 14 |
| Q | 7211 | 2437 | 406 | 286 | 8 | 13 |
| P | 6289 | 3359 | 326 | 215 | 9 | 12 |
| G | 6210 | 3438 | 178 | 91 | 10 | 11 |
| F | 6194 | 3454 | 545 | 349 | 11 | 10 |
| Y | 5943 | 3705 | 1539 | 1066 | 12 | 9 |
| I | 5847 | 3801 | 1136 | 739 | 13 | 8 |
| R | 5523 | 4125 | 1284 | 731 | 14 | 7 |
| T | 5470 | 4178 | 1330 | 835 | 15 | 6 |
| A | 5356 | 4292 | 543 | 274 | 16 | 5 |
| V | 5112 | 4536 | 1809 | 1123 | 17 | 4 |
| S | 5085 | 4563 | 793 | 412 | 18 | 3 |
| K | 3428 | 6220 | 4586 | 2822 | 19 | 2 |
| L | 3197 | 6451 | 3044 | 1442 | 20 | 1 |

## Default R Split Preview

Training/CV peptides contain no `R` anywhere. Evaluation peptides contain `R` at one or more positions.

Train/CV preview:

| Peptide | Fold | Held_Out_AA | Held_Out_Positions | Source_Organism | Species |
| --- | --- | --- | --- | --- | --- |
| AAAAALQAK | 1 | R |  | Homo sapiens | Homo sapiens |
| AAAACTTMK | 2 | R |  | Homo sapiens | Homo sapiens |
| AAAKPTPPK | 3 | R |  | Homo sapiens | Homo sapiens |
| AAAKTLNPK | 4 | R |  | unidentified | unidentified |
| AAAKVILKK | 5 | R |  | Homo sapiens | Homo sapiens |
| AAAPAATAK | 1 | R |  | unidentified | unidentified |
| AAAPALQGK | 2 | R |  | unidentified | unidentified |
| AAAQPLKAK | 3 | R |  | Homo sapiens | Homo sapiens |

Evaluation preview:

| Peptide | Held_Out_AA | Held_Out_Positions | Source_Organism | Species |
| --- | --- | --- | --- | --- |
| AAAPPLSFR | R | 9 | unidentified | unidentified |
| AADFPGIAR | R | 9 | Vibrio cholerae | Vibrio cholerae |
| AAFKRSCLK | R | 5 | Merkel cell polyomavirus | Alphapolyomavirus quintihominis |
| AALDGTFQR | R | 9 | Vaccinia virus WR | Orthopoxvirus vaccinia |
| AAMQRKLEK | R | 5 | SARS coronavirus Tor2 | SARS-CoV1 |
| AAMVLLLRK | R | 8 | Mycobacterium tuberculosis variant bovis BCG str. Pasteur 1173P2 | Mycobacterium tuberculosis |
| AANEIRISK | R | 6 | Salmonella enterica subsp. enterica serovar Typhimurium | Salmonella enterica |
| AAPLRLIAA | R | 5 | Homo sapiens | Homo sapiens |

## Sequence-Logo Style Composition

The files below are information-content stack plots generated from position-wise amino-acid frequencies. The underlying frequency tables are saved as CSVs in `epitope_table_new_data_results/`.

All canonical 9-mers:

![All canonical 9-mer logo](epitope_table_new_data_results/Epitope_Table_Logo_All.svg)

Train/CV peptides with no `R`:

![Train logo](epitope_table_new_data_results/Epitope_Table_Logo_Train_No_R.svg)

Evaluation peptides with `R`:

![Evaluation logo](epitope_table_new_data_results/Epitope_Table_Logo_Evaluation_With_R.svg)

## Downstream Network Training

Run `Epitope_A0301_Heldout_Training_Run.ipynb` after this preparation notebook. That training notebook adds unique label-0 non-binders from `hla_only.txt`, uses HLA-A03:01 negatives first, fills the rest from other alleles to make a 1:1 split, trains AE/one-hot/BLOSUM62 models, and reports CV plus held-out evaluation summaries.